scio-core/src/main/scala/com/spotify/scio/coders/KryoAtomicCoder.scala

```diff
@@ -104,34 +106,156 @@ private[scio] class KryoAtomicCoder[T] extends AtomicCoder[T] {
       kryo.get().writeClassAndObject(output, value)
       output.flush()
     } else {
-      val s = new ByteArrayOutputStream()
-      val output = new Output(s)
+      val os = new BufferedPrefixOutputStream(outStream)
+      val output = new Output(os)
       kryo.get().writeClassAndObject(output, value)
       output.flush()
-
-      VarInt.encode(s.size(), outStream)
-      outStream.write(s.toByteArray)
+      os.finish()
     }
   }
 
   override def decode(inStream: InputStream, context: Context): T = {
     val o = if (context.isWholeStream) {
       kryo.get().readClassAndObject(new Input(inStream))
     } else {
-      val length = VarInt.decodeInt(inStream)
-      if (length < 0) {
-        throw new IOException("invalid length " + length)
-      }
-
-      val value = Array.ofDim[Byte](length)
-      ByteStreams.readFully(inStream, value)
-      kryo.get().readClassAndObject(new Input(value))
+      val is = new BufferedPrefixInputStream(inStream)
+      val obj = kryo.get().readClassAndObject(new Input(is))
+      is.finish()
+      obj
     }
     o.asInstanceOf[T]
   }
 
+  // This method is called by PipelineRunner to sample elements in a PCollection and estimate
+  // size. This could be expensive for collections with small number of very large elements.
+  override def registerByteSizeObserver(value: T, observer: ElementByteSizeObserver,
+                                        context: Context): Unit = value match {
+    // (K, Iterable[V]) is the return type of `groupBy` or `groupByKey`. This could be very slow
+    // when there're few keys with many values.
+    case (key, wrapper: JIterableWrapper[_]) =>
+      observer.update(kryoEncodedElementByteSize(key, Context.OUTER))
+      // FIXME: handle ElementByteSizeObservableIterable[_, _]
+      val i = wrapper.underlying.iterator()
+      while (i.hasNext) {
+        observer.update(kryoEncodedElementByteSize(i.next(), Context.OUTER))
+      }
+    case _ =>
+      observer.update(kryoEncodedElementByteSize(value, context))
+  }
+
+  private def kryoEncodedElementByteSize(obj: Any, context: Context): Long = {
+    val s = new CountingOutputStream(ByteStreams.nullOutputStream())
+    val output = new Output(s)
+    kryo.get().writeClassAndObject(output, obj)
+    output.flush()
+    if (context.isWholeStream) s.getCount else s.getCount + VarInt.getLength(s.getCount)
+  }
+
 }
 
 private[scio] object KryoAtomicCoder {
   def apply[T]: Coder[T] = new KryoAtomicCoder[T]
 }
+
+/**
+ * Buffered output stream that adds length prefix to each buffer block. Useful for large objects
+ * or collections. Based on [[org.apache.beam.sdk.util.BufferedElementCountingOutputStream]].
+ */
+private class BufferedPrefixOutputStream(private val os: OutputStream)
+  extends OutputStream {
+
+  private val buffer = ByteBuffer.allocate(64 * 1024)
```

Comment:
You mean once in input and once in output? I'd have to add an object for it. Again not sure how much value we get since these are not exposed or configurable. It's easier to refactor when the complexity increases and there's a need for it.